dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/blocking/BlockingLogicTask.java

```diff
@@ -0,0 +1,138 @@
+        DependResult expected = taskParameters.getBlockingOpportunity()
+                .equals(BlockingOpportunity.BLOCKING_ON_SUCCESS.getDesc())
+                        ? DependResult.SUCCESS
+                        : DependResult.FAILED;
+        boolean isBlocked = (expected == conditionResult);
+        log.info("blocking opportunity: expected-->{}, actual-->{}", expected, conditionResult);
+        ProcessInstance workflowInstance = processInstanceExecCacheManager
+                .getByProcessInstanceId(taskExecutionContext.getProcessInstanceId()).getProcessInstance();
+        workflowInstance.setBlocked(isBlocked);
+        if (isBlocked) {
+            workflowInstance.setStateWithDesc(WorkflowExecutionStatus.READY_BLOCK, "ready block");
+        }
+        taskExecutionContext.setCurrentExecutionStatus(TaskExecutionStatus.SUCCESS);
+    }
+
+    private DependResult calculateConditionResult() throws MasterTaskExecuteException {
+        // todo: Directly get the task instance from the cache
+        Map<Long, TaskInstance> completeTaskList = taskInstanceDao
+                .findValidTaskListByProcessId(taskExecutionContext.getProcessInstanceId(),
+                        taskExecutionContext.getTestFlag())
+                .stream()
+                .collect(Collectors.toMap(TaskInstance::getTaskCode, Function.identity()));
+
+        // todo: we need to parse the task parameter from TaskExecutionContext
+        TaskInstance taskInstance =
+                processInstanceExecCacheManager.getByProcessInstanceId(taskExecutionContext.getProcessInstanceId())
+                        .getTaskInstance(taskExecutionContext.getTaskInstanceId())
+                        .orElseThrow(() -> new MasterTaskExecuteException("Task instance not found"));
+        DependentParameters dependentParameters = taskInstance.getDependency();
+
+        List<DependResult> tempResultList = new ArrayList<>();
+        for (DependentTaskModel dependentTaskModel : dependentParameters.getDependTaskList()) {
+            List<DependResult> itemDependResult = new ArrayList<>();
+            for (DependentItem item : dependentTaskModel.getDependItemList()) {
+                itemDependResult.add(getDependResultForItem(item, completeTaskList));
+            }
+            DependResult tempResult =
+                    DependentUtils.getDependResultForRelation(dependentTaskModel.getRelation(), itemDependResult);
+            tempResultList.add(tempResult);
+        }
+        return DependentUtils.getDependResultForRelation(dependentParameters.getRelation(), tempResultList);
+    }
+
+    private DependResult getDependResultForItem(DependentItem item, Map<Long, TaskInstance> completeTaskList) {
+
+        DependResult dependResult = DependResult.SUCCESS;
+        if (!completeTaskList.containsKey(item.getDepTaskCode())) {
+            log.info("depend item: {} have not completed yet.", item.getDepTaskCode());
+            dependResult = DependResult.FAILED;
```

Comment:
If not completed, it should be WAITING until dependent finish?

Reply:
Right now, the block task is using dependent pojo... I don't want to refractory these in this pr, we need to remove these logic in the future.